### PR TITLE
Coarse summary of the paper

### PR DESCRIPTION
**Pull request by galeone**
* [X] Please check your author information: `/_data/authors.yml`: if it is not present add it.
* [X] In the article meta the `tags` section is used for the team name. The team list can be find under the section `themeColor` in `_config.yml`: if your team is not present just add it.
* [x] Double check the grammar using Grammarly: https://www.grammarly.com/ (using American English).
* [X] If you can, reduce the code - you're explaining a concept, hence having lots of code makes the article only longer without adding a real information. Short code examples -> detailed description of the concept is the way to go.
* [X] Keep the article formal - reduce/remove the slang expressions and the punctuations like `...` if not strictly necessary.
* [X] Use the oxford comma (setting up Grammarly to use American English should forces you to use it)
* [X] Do not hotlink images: put the images inside a subfolder in the images folder of the blog and link there. Double check the images copyright before using them.
* [X] The ideal article should start from a problem -> the article content should start from it and describe the flow to solve it.

Still a draft - improve and publish this when we are ready for releasing the tf2 implementation.
In the meantime, we can improve this article (that's just a summary with several parts extracted from the papery).

### CODE REVIEW

File: _posts/2021-04-21-3d-bonet.md [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+The global feature vector $$\textbf{F}_g$$ is fed through 2 fully connected layers with Leaky ReLU, followed by 2 parallel FC layers.
+
+- 1 layer outputs a 6H dimensional vector, reshaped into a $$ H \times 2 \times 3$$ tensor, where H is the **predefined and fixes number of bounding boxes**
+- 1 layer outputs a H dimensional vector followed by a **sigmoid** to represent the bounding box scores.
+
+
+**Bounding Box Association Layer**
+
+Given the H predicted bboxes $$ \textbf{B} \in \mathbb{R}^{H \times 2 \times 3} $$ how to use the gt bboxes $\textbf{\bar{B}} \in \mathbb{R}^{T \times 2 \times 3}$$ to supervise the network?
```
Comment (marcov868) [inline _posts/2021-04-21-3d-bonet.md]: At the beginning of the second formula, a `$` is missing.
Moreover, I don't know if I have some issues with my configuration, but the second formula doesn't display correctly, probably the `\bar` goes before the `\textbf` :
` $$ \bar{\textbf{B}} \in \mathbb{R}^{T \times 2 \times 3}  $$ `

File: _posts/2021-04-21-3d-bonet.md [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+**Optimal Association Formulation**
+
+Let $$\textbf{A}$$ be a boolean association matrix where $$\textbf{A}_{i,j} = 1 \iff \text{the} \quad i^{\text{th}}$$ predicted bbox is assigned to the $$j^{\text{th}}$$ gt box.
+
+$$\textbf{A}$$ = the association index.
+
+Let $$\textbf{C}$$ be the association **cost matrix** where $$\textbf{C}_{i,j}$$ represents the cost that the $$i^{th}$$ predicted box is assigned to the $$j^{\text{th}}$$ gt box.
+
+The less the cost, the more similar are 2 boxes. Hence, the box association problem is to find the optimal assignment matrix $$	extbf{A} with the **minimal cost overall**:
```
Comment (marcov868) [inline _posts/2021-04-21-3d-bonet.md]: There's a typo in `$$\textbf{A}$$`

**Comment by marcov868** [comment]
I read the paper just one time but I found the summary very clear, it follows the skeleton of the article summarizing the main concepts with the addition of some useful notes. 
I think could be clear also for someone that didn't read the paper.

**Comment by galeone** [comment]
> I read the paper just one time but I found the summary very clear, it follows the skeleton of the article summarizing the main concepts with the addition of some useful notes.
> I think could be clear also for someone that didn't read the paper.

:heart:

File: _posts/2021-10-13-3d-bonet.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+title:  "3D instance segmentation with 3D-BoNet: analysis and TensorFlow 2 implementation"
+slug:   "3D instance segmentation with 3D-BoNet: analysis and TensorFlow 2 implementation"
+date:    2021-10-13 8:00:00
+categories: machine-learning
+image: images/3dbonet/3d-bonet-architecture.png
+tags: machine-learning
+---
+
+Instance segmentation on point clouds is a challenging problem. Point clouds are unordered, sparse, non-uniform, and finding non-regulars shapes is not trivial. Moreover, the number of points in a scene can easily surpass the million, making every algorithm on point clouds very computationally expensive.
+The 3D-BoNet architecture [1] has been introduced recently to tackle the problem. This network is trainable end-to-end, and it mixes the results of a bounding box prediction branch with the output of an instance segmentation branch. The article contains a walkthrough of the paper, section by section, and at the end introduces the steps we followed to convert the original implementation (in TensorFlow 1) to our own implementation in TensorFlow 2.
```
Comment (marcov868) [inline _posts/2021-10-13-3d-bonet.md]:
```suggestion
The 3D-BoNet architecture [1] has been introduced recently to tackle the problem. This network is trainable end-to-end, and it mixes the results of a bounding box prediction branch with the output of an instance segmentation branch. The article contains a walkthrough of the paper, section by section, and eventually introduces the steps we followed to convert the original implementation (in TensorFlow 1) to our implementation in TensorFlow 2.
```

File: _posts/2021-10-13-3d-bonet.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+Analysis of point clouds is burdensome because point clouds are inherently unordered, unstructured, and non-uniform. The approaches to work with point clouds can be divided into three families: projection-based, discretization-based, and point-based methods. In project-based methods, the point cloud is projected to a plane, and the resulting image is analyzed with image processing techniques. Discretization-based approaches organize points in a grid of voxels summarizing the point features in voxels. This discretization allows using 3D convolutional neural networks because it imposes a regular geometric structure. The main drawback of this representation is its memory footprint, which can be very high if the scene covers a large section of space.
+
+The 3D-BoNet approach is point-based like SGPN [2], ASIS [3], JSIS3D [4], MASC [5], 3D-BEVIS [6], but all these methods do not explicitly detect object boundaries (the bounding boxes). Furthermore, they require computationally intensive post-processing, like mean-shift clustering.
```
Comment (marcov868) [inline _posts/2021-10-13-3d-bonet.md]:
```suggestion
The 3D-BoNet approach is point-based like SGPN [2], ASIS [3], JSIS3D [4], MASC [5], 3D-BEVIS [6], but all these methods do not explicitly detect object boundaries (i.e, the bounding boxes). Furthermore, they require computationally intensive post-processing, such as mean-shift clustering.
```

File: _posts/2021-10-13-3d-bonet.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+Differently from previous models, 3D-BoNet directly regresses the vertices of instance bounding boxes. Then it selects the points belonging to the instance with a binary classifier applied to the points within the bounding box. For doing this, the authors introduced a **bounding box prediction module** and a series of carefully designed loss functions to directly learn object boolean masks.
```
Comment (marcov868) [inline _posts/2021-10-13-3d-bonet.md]:
```suggestion
Unlike previous models, 3D-BoNet directly regresses the vertices of instance bounding boxes. Then, it selects the points belonging to the instance applying a binary classifier to the points within the bounding box. For doing this, the authors introduced a **bounding box prediction module** and a series of carefully designed loss functions to directly learn object boolean masks.
```

File: _posts/2021-10-13-3d-bonet.md [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+The pure 3D-BoNet method does not require post-processing, but it only works on small point clouds (usually composed of 4096 points). It is necessary to split the cloud into spatial blocks to manage more numerous point clouds. A spatial block contains the points belonging to a vertical column of space with a basis measuring a 1-meter square. The number of points in a block is constant to allow the creation of batches. For this reason, from a single spatial block, many processing blocks are created by sampling points.
```
Comment (marcov868) [inline _posts/2021-10-13-3d-bonet.md]:
```suggestion
The pure 3D-BoNet method does not require post-processing, but it only works on small point clouds, usually composed of 4096 points. It is necessary to split the cloud into spatial blocks to manage more numerous point clouds. A spatial block contains the points belonging to a vertical column of space with a basis measuring a 1-meter square. The number of points in a block is constant to allow the creation of batches. For this reason, from a single spatial block, many processing blocks are created by sampling points. 
```

File: _posts/2021-10-13-3d-bonet.md [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+</div>
+
+The architecture is very similar to a traditional architecture used on images: backbone (feature extractor) and different heads for multi-task learning.
+
+The input point cloud $$\textbf{P} \in \mathbb{R}^{N \times k_0}$$ has $$N$$ points and $$ k_0 $$ feature channels (such as the location coordinates $$\{x,y,z\}$$ and color $$\{r,g,b\}$$). The backbone network extracts **point local features** $$\textbf{F}_l \in \mathbb{R}^{N \times k}$$, where $$k$$ is the length of feature vectors.
+Then, local features are aggregated in a **global feature vector** $$\textbf{F}_g \in \mathbb{R}^{1 \times k}$$.
+
+The **Bounding Box Prediction Branch** (BBPB) takes global features $$\textbf{F}_g$$ as input and regresses a **predefined and fixed set of bounding boxes** $$B$$ and the corresponding box scores $$B_s$$.
+
+During the training, this branch is supervised with ground truth boxes. A **bounding box association layer** associates the most similar predicted bounding box to every ground truth box. The output of the association layer is a list of indices $$\textbf{A}$$ that describes the best pairing between boxes. Paired boxes are then compared with a multi-criteria loss.
```
Comment (marcov868) [inline _posts/2021-10-13-3d-bonet.md]:
```suggestion
During the training, this branch is supervised with ground truth boxes. The **bounding box association layer** associates the most similar predicted bounding box to every ground truth box. The output of the association layer is a list of indices $$\textbf{A}$$ that describes the best pairing between boxes. Paired boxes are then compared with a multi-criteria loss.
```

File: _posts/2021-10-13-3d-bonet.md [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+#### Bounding Box Association Layer
+
+Given $$H$$ predicted bounding boxes $$\textbf{B} \in \mathbb{R}^{H \times 2 \times 3} $$ how can we use ground truth bounding boxes $$\bar{\textbf{B}} \in \mathbb{R}^{T \times 2 \times 3}$$ to supervise the network?
+
+The Bounding Box Association Layer associates to every $$T$$ ground truth box one of the $$H$$ predicted boxes when $$T \leq H$$. When $$T > H$$, the largest $$H$$ ground truth boxes only are retained.
+
+The associations between boxes can be encoded in a boolean **association matrix** $$A \in \mathbb{R}^{H \times T}$$ such that $$\textbf{A}_{i,j} = 1$$ if and only if the $$i^{\text{th}}$$ predicted bbox is assigned to the $$j^{\text{th}}$$ ground truth box.
+At the same time the dissimilarity between any pair of boxes can be stored in the **cost matrix** $$\textbf{C} \in \mathbb{R}^{H \times T}$$
```
Comment (marcov868) [inline _posts/2021-10-13-3d-bonet.md]:
```suggestion
At the same time the dissimilarity between any pair of boxes can be stored in the **cost matrix** $$\textbf{C} \in \mathbb{R}^{H \times T}$$.
```

File: _posts/2021-10-13-3d-bonet.md [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+The final association cost between the $$i^\text{th}$$ predicted bounding box and the $$j^\text{th}$$ ground truth box is defined as:
+
+$$ C_{i,j} = C^{ed}_{i,j} + C^{\text{sIoU}}_{i,j} + C^{\text{ces}}_{i,j}.$$
+
+The entries in the cost matrix $$\mathbf{C}$$ are filled according to this formula and the association matrix $$ \mathbf{A} $$ is computed through the Hungarian algorithm.
+
+#### Loss Functions
+
+The association matrix $$ \mathbf{A} $$ allows to reorder the predicted boxes placing them in correspondence of the most similar ground truth box.
```
Comment (marcov868) [inline _posts/2021-10-13-3d-bonet.md]:
```suggestion
The association matrix $$ \mathbf{A} $$ allows reordering the predicted boxes, placing them in correspondence with the most similar ground truth box.
```

File: _posts/2021-10-13-3d-bonet.md [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+The loss function that supervises the model is composed of two parts.
+
+**Multi-criteria Loss for Box prediction**.  We reuse and minimize the association cost previously defined, i.e.
+
+$$ \mathcal{l_\text{bbox}} = \frac{1}{T} \sum_{t=1}^{T}{C^{ed}_{t,t} + C^{\text{sIoU}}_{t,t} + C^{\text{ces}}_{t,t}}.$$
+
+Here we only minimize the cost of $$T$$ paired boxes, while the remaining $$H-T$$ boxes are ignored.
+In practice, $$ \mathcal{l_\text{bbox}} $$ corresponds to the minimal cost found by the Hungarian algorithm.
+
+**Loss for Box Score Prediction**. The predicted box scores $B_s$ aim to indicate the validity of the corresponding predicted boxes.
```
Comment (marcov868) [inline _posts/2021-10-13-3d-bonet.md]:
```suggestion
**Loss for Box Score Prediction**. The predicted box scores $$B_s$$ aim to indicate the validity of the corresponding predicted boxes. 
```

File: _posts/2021-10-13-3d-bonet.md [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+The predicted instance masks $$M$$ are associated with the ground truth masks according to the association matrix $$\mathbf{A}$$. Due to the imbalance of instance and background points, the **focal loss** with default hyper-parameters is used instead of standard cross-entropy loss. As in the bounding box prediction branch, only the valid $$T$$ paired masks are used for the loss $$\mathcal{l_\text{pmask}}$$.
+
+### Semantic Segmentation and Instance Segmentation
+
+The 3D-BoNet architecture can be adapted to perform semantic segmentation along with object detection. 
+The semantic segmentation task is a pointwise prediction task like the point mask prediction task.
+Thus, the semantic segmentation Branch is very similar to the point mask prediction branch.
```
Comment (marcov868) [inline _posts/2021-10-13-3d-bonet.md]:
```suggestion
Thus, the semantic segmentation branch is very similar to the point mask prediction branch.
```

File: _posts/2021-10-13-3d-bonet.md [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+The problem is even more relevant during training when gradients have to be stored along with layer outputs.
+For this reason, it is necessary to devise a strategy to train and use the model on large point clouds.
+
+### Subsampling
+
+First of all, it is always a good idea to subsample the point cloud to remove points in highly dense regions.
+There are many methods to prune points in dense regions: in our implementation, the subsampling preserves only one point per voxel in a grid.
+
+First of all, it is always a good idea to subsample the point cloud to remove points in highly dense regions.
+There are many methods to prune points in dense regions: in our implementation, the subsampling preserves only one point per voxel in a grid.
```
Comment (marcov868) [inline _posts/2021-10-13-3d-bonet.md]: > First of all, it is always a good idea to subsample the point cloud to remove points in highly dense regions.
> There are many methods to prune points in dense regions: in our implementation, the subsampling preserves only one point per voxel in a grid.

These two sentences are repeated

File: _posts/2021-10-13-3d-bonet.md [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+More in detail, the 3D space of points is divided virtually into voxels by a cubic grid whose size is a parameter of the experimenter.
```
Comment (marcov868) [inline _posts/2021-10-13-3d-bonet.md]:
```suggestion
More in detail, the 3D space of points is divided virtually into voxels by a cubic grid whose size is a parameter of the experiment.
```

File: _posts/2021-10-13-3d-bonet.md [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+For every voxel, a single point is preserved. Its coordinates are obtained by averaging the coordinates of all points in the voxel.
+
+This method is easy to implement, and it can reduce the number of points in the point cloud by one order of magnitude, but it is still not sufficient.
+
+### Block Merging
+
+The authors of 3D-BoNet propose to use Block Merging [2] to split the point cloud into fixed-sized blocks of points. Let us see how blocks are created.
+First of all,  the 3D space is partitioned in overlapping squared columns of infinite height. 
+For example, for a dataset composed of rooms, the squared base of every column usually measures $$1 \text{m}^2$$,
+and the overlap between consecutive columns is usually $0.5 \text{m}$$ in both directions.
```
Comment (marcov868) [inline _posts/2021-10-13-3d-bonet.md]:
```suggestion
and the overlap between consecutive columns is usually $$0.5 \text{m}$$ in both directions.
```

File: _posts/2021-10-13-3d-bonet.md [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+## Implementation
+
+The 3D-BoNet framework can be used in combination with any backbone: in our experiments, we implemented and used PointNet or PointNet++.
+Our implementation is directly inspired by the [official implementation](https://github.com/Yang7879/3D-BoNet), but it has many advantages over the original code.
+
+Our implementation:
+- is compatible with TensorFlow 2, while the original code support TensorFlow 1 only;
+- is organized into commented functions and classes;
+- reduces the number of custom operations on point cloud from 3 to just 1, replacing custom operations that have to be compiled manually with native TensorFlow functions;
+- implements the BlockMerging method in TensorFlow, allowing to export it with the model in the SavedModel format simplifying the deployment of the model
```
Comment (marcov868) [inline _posts/2021-10-13-3d-bonet.md]:
```suggestion
- implements the Block Merging method in TensorFlow, allowing to export it with the model in the SavedModel format simplifying the deployment of the model
```

File: _posts/2021-10-13-3d-bonet.md [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+The S3DIS dataset is the only dataset compatible with the original code released, and, in addition, the authors of the 3D-BoNet paper have published the dataset in its processed form with the point clouds already split into blocks. The reproduction of the results on S3DIS is thus not very hard, but it also hides most of the challenges posed by Block Merging.
+
+We have encountered much more problems while working with the ScanNet dataset. This dataset is composed of point clouds larger than the ones in S3DIS, requiring a heavy subsampling. Furthermore, the division into blocks has to be performed from scratch.
+The authors declare great detection results on ScanNet without releasing the necessary code to reproduce them, but we have not attained the stated results. Even after numerous experiments that cover a wide range of hyper-parameters for subsampling and Block Merging, our trained models obtain very low accuracy.
+This fact raises some doubt about the effectiveness of the 3D-BoNet framework for the segmentation of large point clouds.
+
+You can find our implementation on Github: https://github.com/zurutech/3DBoNet2
```
Comment (marcov868) [inline _posts/2021-10-13-3d-bonet.md]:
```suggestion
You can find our implementation on [Github](https://github.com/zurutech/3DBoNet2).
```

Comment (lucagrementieri) [reply _posts/2021-10-13-3d-bonet.md]: I like to show the explicit link.

Comment (marcov868) [reply _posts/2021-10-13-3d-bonet.md]: Perhaps I would only make it clickable

Comment (lucagrementieri) [reply _posts/2021-10-13-3d-bonet.md]: Ah sure! I thought that links were automatically clickable.

Comment (marcov868) [reply _posts/2021-10-13-3d-bonet.md]: It makes sense, probably it's a problem of my local settings

File: _posts/2021-10-13-3d-bonet.md (hunk 3 above)
Comment (lucagrementieri) [reply _posts/2021-10-13-3d-bonet.md]: I have changed the sentence  applying the suggested edits too.

File: _posts/2021-10-13-3d-bonet.md (hunk 16 above)
Comment (lucagrementieri) [reply _posts/2021-10-13-3d-bonet.md]: I have made the link explicit.

File: _posts/2021-10-13-3d-bonet.md [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+Instance segmentation aims to classify every point like semantic segmentation, but now distinct objects from the same semantic class have different labels. For this reason, an instance segmentation model should be able to reason over multiple scales: it has to recognize fine details to classify the points correctly, and, at the same time, it has to take into account objects in their entirety.
+
+Analysis of point clouds is burdensome because point clouds are inherently unordered, unstructured, and non-uniform. The approaches to work with point clouds can be divided into three families: projection-based, discretization-based, and point-based methods. In project-based methods, the point cloud is projected to a plane, and the resulting image is analyzed with image processing techniques. Discretization-based approaches organize points in a grid of voxels summarizing the point features in voxels. This discretization allows using 3D convolutional neural networks because it imposes a regular geometric structure. The main drawback of this representation is its memory footprint, which can be very high if the scene covers a large section of space.
```
Comment (galeone) [inline _posts/2021-10-13-3d-bonet.md]: project-based should be projection-based as mentioned in the previous statement

File: _posts/2021-10-13-3d-bonet.md [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+
+The 3D-BoNet approach is point-based like SGPN [2], ASIS [3], JSIS3D [4], MASC [5], 3D-BEVIS [6], but all these methods do not explicitly detect object boundaries (i.e, the bounding boxes). Furthermore, they require computationally intensive post-processing, such as mean-shift clustering.
+
+Unlike previous models, 3D-BoNet directly regresses the vertices of instance bounding boxes. Then, it selects the points belonging to the instance applying a binary classifier to the points within the bounding box. For doing this, the authors introduced a **bounding box prediction module** and a series of carefully designed loss functions to directly learn object boolean masks.
+
+The pure 3D-BoNet method does not require post-processing, but it only works on small point clouds, usually composed of 4096 points. It is necessary to split the cloud into spatial blocks to manage more numerous point clouds. A spatial block contains the points belonging to a vertical square column of space. The number of points in a block is constant to allow the creation of batches. For this reason, from a single spatial block, many processing blocks are created by sampling points.
```
Comment (galeone) [inline _posts/2021-10-13-3d-bonet.md]: "A spatial block contains the points belonging to a vertical square column of space."

Sounds strange. I would rephrase it like "A spatial block contains the point belonging to a rectangular parallelepiped with square base" or something similar (?)

File: _posts/2021-10-13-3d-bonet.md [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+### Bounding Box Prediction Branch (BBPB)
+
+<div class="blog-image-container">
+    <figure>
+        <img class="blog-image" alt="bbox prediction" src="/images/3dbonet/3d-bo-net-bbox.png">
+    <figcaption> Bounding Box Predictiond Branch architecture </figcaption>
+    </figure>
+</div>
+
+The goal of the BBPB is to predict a rectangular bounding box for each instance in a single forward pass without relying on predefined spatial anchors or RPN (Region Proposal Network). A bounding box is parametrized by a pair of opposed vertices: $$ \{ [x_\text{min}, y_\text{min}, z_\text{min}], [x_\text{max}, y_\text{max}, z_\text{max}]\} $$
```
Comment (galeone) [inline _posts/2021-10-13-3d-bonet.md]: "rectangular bounding box" is somehow redundant since a bounding box is the minimum enclosing rectangle, hence I'd remove "rectangular"

File: _posts/2021-10-13-3d-bonet.md [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+There are many methods to prune points in dense regions: in our implementation, the subsampling preserves only one point per voxel in a grid.
+More in detail, the 3D space of points is divided virtually into voxels by a cubic grid whose size is a parameter of the experiment.
+For every voxel, a single point is preserved. Its coordinates are obtained by averaging the coordinates of all points in the voxel.
+
+This method is easy to implement, and it can reduce the number of points in the point cloud by one order of magnitude, but it is still not sufficient.
+
+### Block Merging
+
+The authors of 3D-BoNet propose to use Block Merging [2] to split the point cloud into fixed-sized blocks of points. Let us see how blocks are created.
+First of all,  the 3D space is partitioned in overlapping squared columns of infinite height.
```
Comment (galeone) [inline _posts/2021-10-13-3d-bonet.md]: squared columns ->  rectangular square-based parallelepipeds?

File: _posts/2021-10-13-3d-bonet.md [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+title:  "3D instance segmentation with 3D-BoNet: analysis and TensorFlow 2 implementation"
+slug:   "3D instance segmentation with 3D-BoNet: analysis and TensorFlow 2 implementation"
+date:    2021-10-13 8:00:00
+categories: machine-learning
+image: images/3dbonet/3d-bonet-architecture.png
+tags: machine-learning
+---
+
+Instance segmentation on point clouds is a challenging problem. Point clouds are unordered, sparse, non-uniform, and finding non-regulars shapes is not trivial. Moreover, the number of points in a scene can easily surpass the million, making every algorithm on point clouds very computationally expensive.
+The 3D-BoNet architecture [1] has been introduced recently to tackle the problem. This network is trainable end-to-end, and it mixes the results of a bounding box prediction branch with the output of a semantic segmentation branch to perform instance segmentation. The article contains a walkthrough of the paper, section by section, and eventually introduces our implementation in TensorFlow 2 obtained refactoring the original code in TensorFlow 1.
```
Comment (galeone) [inline _posts/2021-10-13-3d-bonet.md]: I would add another line like

```
If you're not interested in the walkthrough of the paper, you can directly jump a the [implementation](#implementation).
```
